File: mmupnp/src/main/java/net/mm2d/upnp/internal/server/SsdpServerDelegate.java
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.server;

import net.mm2d.log.Logger;
import net.mm2d.upnp.Http;
import net.mm2d.upnp.SsdpMessage;
import net.mm2d.upnp.internal.thread.TaskExecutors;
import net.mm2d.upnp.internal.util.IoUtils;
import net.mm2d.upnp.util.NetworkUtils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.Inet4Address;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.InterfaceAddress;
import java.net.MalformedURLException;
import java.net.MulticastSocket;
import java.net.NetworkInterface;
import java.net.SocketTimeoutException;
import java.net.URL;
import java.net.UnknownHostException;
import java.util.List;
import java.util.concurrent.FutureTask;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * SsdpServerの共通処理を実装するクラス。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
class SsdpServerDelegate implements SsdpServer, Runnable {
    interface Receiver {
        /**
         * メッセージ受信後の処理、サブクラスにより実装する。
         *
         * @param sourceAddress 送信元アドレス
         * @param data          受信したデータ
         * @param length        受信したデータの長さ
         */
        void onReceive(
                @Nonnull InetAddress sourceAddress,
                @Nonnull byte[] data,
                int length);
    }

    private static final long PREPARE_TIMEOUT = 3000L;

    @Nonnull
    private final TaskExecutors mTaskExecutors;
    @Nonnull
    private final InterfaceAddress mInterfaceAddress;
    @Nonnull
    private final Receiver mReceiver;
    @Nonnull
    private final Address mAddress;
    @Nonnull
    private final NetworkInterface mInterface;

    private final int mBindPort;
    @Nullable
    private FutureTask<?> mFutureTask;

    private volatile boolean mReady;
    @Nullable
    private MulticastSocket mSocket;

    /**
     * 使用するインターフェースを指定してインスタンス作成。
     *
     * <p>使用するポートは自動割当となる。
     *
     * @param receiver         パケット受信時にコールされるreceiver
     * @param networkInterface 使用するインターフェース
     * @param address          モード
     */
    SsdpServerDelegate(
            @Nonnull final TaskExecutors executors,
            @Nonnull final Receiver receiver,
            @Nonnull final Address address,
            @Nonnull final NetworkInterface networkInterface) {
        this(executors, receiver, address, networkInterface, 0);
    }

    /**
     * 使用するインターフェースとポート指定してインスタンス作成。
     *
     * @param receiver         パケット受信時にコールされるreceiver
     * @param networkInterface 使用するインターフェース
     * @param bindPort         使用するポート
     * @param address          モード
     */
    SsdpServerDelegate(
            @Nonnull final TaskExecutors executors,
            @Nonnull final Receiver receiver,
            @Nonnull final Address address,
            @Nonnull final NetworkInterface networkInterface,
            final int bindPort) {
        mInterfaceAddress = address == Address.IP_V4 ?
                findInet4Address(networkInterface.getInterfaceAddresses()) :
                findInet6Address(networkInterface.getInterfaceAddresses());
        mTaskExecutors = executors;
        mReceiver = receiver;
        mAddress = address;
        mInterface = networkInterface;
        mBindPort = bindPort;
    }

    /**
     * SsdpMessageのLocationに正常なURLが記述されており、
     * 記述のアドレスとパケットの送信元アドレスに不一致がないか検査する。
     *
     * @param message       確認するSsdpMessage
     * @param sourceAddress 送信元アドレス
     * @return true:送信元との不一致を含めてLocationに不正がある場合。false:それ以外
     */
    public static boolean isInvalidLocation(
            @Nonnull final SsdpMessage message,
            @Nonnull final InetAddress sourceAddress) {
        return !isValidLocation(message, sourceAddress);
    }

    private static boolean isValidLocation(
            @Nonnull final SsdpMessage message,
            @Nonnull final InetAddress sourceAddress) {
        final String location = message.getLocation();
        if (!Http.isHttpUrl(location)) {
            return false;
        }
        try {
            final InetAddress locationAddress = InetAddress.getByName(new URL(location).getHost());
            return sourceAddress.equals(locationAddress);
        } catch (final MalformedURLException | UnknownHostException ignored) {
        }
        return false;
    }

    /**
     * マルチキャストアドレスを返す。
     *
     * @return マルチキャストアドレス
     */
    @Nonnull
    Address getAddress() {
        return mAddress;
    }

    /**
     * SSDPに使用するアドレス。
     *
     * @return SSDPで使用するInetAddress
     */
    @Nonnull
    InetAddress getSsdpInetAddress() {
        return mAddress.getInetAddress();
    }

    /**
     * SSDPに使用するアドレス＋ポートの文字列。
     *
     * @return SSDPに使用するアドレス＋ポートの文字列。
     */
    @Nonnull
    String getSsdpAddressString() {
        return mAddress.getAddressString();
    }

    // VisibleForTesting
    @Nonnull
    static InterfaceAddress findInet4Address(@Nonnull final List<InterfaceAddress> addressList) {
        for (final InterfaceAddress address : addressList) {
            if (address.getAddress() instanceof Inet4Address) {
                return address;
            }
        }
        throw new IllegalArgumentException("ni does not have IPv4 address.");
    }

    // VisibleForTesting
    @Nonnull
    static InterfaceAddress findInet6Address(@Nonnull final List<InterfaceAddress> addressList) {
        for (final InterfaceAddress address : addressList) {
            final InetAddress inetAddress = address.getAddress();
            if (inetAddress instanceof Inet6Address) {
                if (inetAddress.isLinkLocalAddress()) {
                    return address;
                }
            }
        }
        throw new IllegalArgumentException("ni does not have IPv6 address.");
    }

    @Nonnull
    public InterfaceAddress getInterfaceAddress() {
        return mInterfaceAddress;
    }

    @Nonnull
    public InetAddress getLocalAddress() {
        return mInterfaceAddress.getAddress();
    }

    @SuppressWarnings("Duplicates")
    @Override
    public void start() {
        if (mFutureTask != null) {
            stop();
        }
        mReady = false;
        mFutureTask = new FutureTask<>(this, null);
        mTaskExecutors.server(mFutureTask);
    }

    @SuppressWarnings("Duplicates")
    @Override
    public void stop() {
        final FutureTask<?> task = mFutureTask;
        if (task == null) {
            return;
        }
        task.cancel(false);
        mFutureTask = null;
        IoUtils.closeQuietly(mSocket);
    }

    @Override
    public void send(@Nonnull final SsdpMessage message) {
        mTaskExecutors.io(() -> sendInner(message));
    }

    private void sendInner(@Nonnull final SsdpMessage message) {
        if (!waitReady()) {
            Logger.w("socket is not ready");
            return;
        }
        final MulticastSocket socket = mSocket;
        if (socket == null) {
            return;
        }
        Logger.d(() -> "send from " + mInterfaceAddress + ":\n" + message);
        try {
            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
            message.writeData(baos);
            final byte[] data = baos.toByteArray();
            socket.send(new DatagramPacket(data, data.length, mAddress.getSocketAddress()));
        } catch (final IOException e) {
            Logger.w(e);
        }
    }

    @SuppressWarnings("Duplicates")
    private synchronized boolean waitReady() {
        final FutureTask<?> task = mFutureTask;
        if (task == null || task.isDone()) {
            return false;
        }
        if (!mReady) {
            try {
                wait(PREPARE_TIMEOUT);
            } catch (final InterruptedException ignored) {
            }
        }
        return mReady;
    }

    private synchronized void notifyReady() {
        mReady = true;
        notifyAll();
    }

    @Override
    public void run() {
        final String suffix = (mBindPort == 0 ? "-ssdp-notify-" : "-ssdp-search-")
                + mInterface.getName() + "-"
                + NetworkUtils.toSimpleString(mInterfaceAddress.getAddress());
        final Thread thread = Thread.currentThread();
        thread.setName(thread.getName() + suffix);
        if (mFutureTask == null || mFutureTask.isCancelled()) {
            return;
        }
        try {
            final MulticastSocket socket = createMulticastSocket(mBindPort);
            mSocket = socket;
            if (mBindPort != 0) {
                socket.joinGroup(getSsdpInetAddress());
            }
            notifyReady();
            receiveLoop(socket);
        } catch (final IOException ignored) {
        } finally {
            final MulticastSocket socket = mSocket;
            if (mBindPort != 0 && socket != null) {
                try {
                    socket.leaveGroup(getSsdpInetAddress());
                } catch (final IOException ignored) {
                }
            }
            IoUtils.closeQuietly(mSocket);
            mSocket = null;
        }
    }

    private boolean isCancelled() {
        final FutureTask<?> task = mFutureTask;
        return task == null || task.isCancelled();
    }

    /**
     * 受信処理を行う。
     *
     * @throws IOException 入出力処理で例外発生
     */
    // VisibleForTesting
    void receiveLoop(@Nonnull final MulticastSocket socket) throws IOException {
        final byte[] buf = new byte[1500];
        while (!isCancelled()) {
            try {
                final DatagramPacket dp = new DatagramPacket(buf, buf.length);
                socket.receive(dp);
                if (isCancelled()) {
                    break;
                }
                mReceiver.onReceive(dp.getAddress(), dp.getData(), dp.getLength());
            } catch (final SocketTimeoutException ignored) {
            }
        }
    }

    // VisibleForTesting
    @Nonnull
    MulticastSocket createMulticastSocket(final int port) throws IOException {
        final MulticastSocket socket = new MulticastSocket(port);
        socket.setNetworkInterface(mInterface);
        socket.setTimeToLive(4);
        return socket;
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/internal/server/SsdpServerDelegate.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.server

import net.mm2d.log.Logger
import net.mm2d.upnp.Http
import net.mm2d.upnp.SsdpMessage
import net.mm2d.upnp.internal.thread.TaskExecutors
import net.mm2d.upnp.internal.util.closeQuietly
import net.mm2d.upnp.util.findInet4Address
import net.mm2d.upnp.util.findInet6Address
import net.mm2d.upnp.util.toSimpleString
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.*
import java.util.concurrent.FutureTask
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * A class that implements the common part of [SsdpServer].
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 *
 * @param taskExecutors taskExecutors
 * @param address Multicast address
 * @param networkInterface network interface
 * @param bindPort port number
 */
internal class SsdpServerDelegate(
    private val taskExecutors: TaskExecutors,
    val address: Address,
    private val networkInterface: NetworkInterface,
    private val bindPort: Int = 0
) : SsdpServer, Runnable {
    val interfaceAddress: InterfaceAddress =
        if (address == Address.IP_V4)
            networkInterface.findInet4Address()
        else
            networkInterface.findInet6Address()
    private var socket: MulticastSocket? = null
    private var receiver: ((sourceAddress: InetAddress, data: ByteArray, length: Int) -> Unit)? = null

    private var futureTask: FutureTask<*>? = null
    private val lock = ReentrantLock()
    private val condition = lock.newCondition()
    private var ready = false

    fun setReceiver(receiver: ((sourceAddress: InetAddress, data: ByteArray, length: Int) -> Unit)?) {
        this.receiver = receiver
    }

    fun getSsdpInetAddress(): InetAddress = address.ssdpInetAddress

    fun getSsdpAddressString(): String = address.ssdpAddressString

    fun getLocalAddress(): InetAddress = interfaceAddress.address

    // VisibleForTesting
    @Throws(IOException::class)
    internal fun createMulticastSocket(port: Int): MulticastSocket {
        return MulticastSocket(port).also {
            it.networkInterface = networkInterface
            it.timeToLive = 4
        }
    }

    override fun start() {
        receiver ?: throw IllegalStateException("receiver must be set")
        futureTask ?: stop()
        lock.withLock {
            ready = false
            FutureTask(this, null).also {
                futureTask = it
                taskExecutors.server(it)
            }
        }
    }

    override fun stop() {
        lock.withLock {
            futureTask?.cancel(false)
            futureTask = null
            socket.closeQuietly()
        }
    }

    override fun send(messageSupplier: () -> SsdpMessage) {
        taskExecutors.io { sendInner(messageSupplier()) }
    }

    private fun sendInner(message: SsdpMessage) {
        if (!waitReady()) {
            Logger.w("socket is not ready")
            return
        }
        val socket = socket ?: return
        Logger.d { "send from $interfaceAddress:\n$message" }
        try {
            val data = ByteArrayOutputStream().also {
                message.writeData(it)
            }.toByteArray()
            socket.send(DatagramPacket(data, data.size, address.ssdpSocketAddress))
        } catch (e: IOException) {
            Logger.w(e)
        }
    }

    private fun waitReady(): Boolean = lock.withLock {
        val task = futureTask ?: return false
        if (task.isDone) return false
        if (!ready) {
            try {
                condition.awaitNanos(PREPARE_TIMEOUT_NANOS)
            } catch (ignored: InterruptedException) {
            }
        }
        ready
    }

    private fun notifyReady() {
        lock.withLock {
            ready = true
            condition.signalAll()
        }
    }

    // VisibleForTesting
    internal fun isCanceled(): Boolean = futureTask?.isCancelled ?: true

    override fun run() {
        val suffix = (if (bindPort == 0) "-ssdp-notify-" else "-ssdp-search-") +
            networkInterface.name + "-" + interfaceAddress.address.toSimpleString()
        Thread.currentThread().let {
            it.name = it.name + suffix
        }
        if (isCanceled()) {
            return
        }
        try {
            val socket = createMulticastSocket(bindPort)
            this.socket = socket
            if (bindPort != 0) {
                socket.joinGroup(getSsdpInetAddress())
            }
            notifyReady()
            receiveLoop(socket)
        } catch (ignored: IOException) {
        } finally {
            if (bindPort != 0) {
                socket?.leaveGroup(getSsdpInetAddress())
            }
            socket.closeQuietly()
            socket = null
        }
    }

    // VisibleForTesting
    @Throws(IOException::class)
    internal fun receiveLoop(socket: MulticastSocket) {
        val buf = ByteArray(1500)
        while (!isCanceled()) {
            try {
                val dp = DatagramPacket(buf, buf.size)
                socket.receive(dp)
                if (isCanceled()) {
                    break
                }
                receiver?.invoke(dp.address, dp.data, dp.length)
            } catch (ignored: SocketTimeoutException) {
            }
        }
    }

    companion object {
        private val PREPARE_TIMEOUT_NANOS = TimeUnit.SECONDS.toNanos(3)
    }
}

internal val DEFAULT_SSDP_MESSAGE_FILTER: (SsdpMessage) -> Boolean = { true }

/**
 * A normal URL is described in the Location of SsdpMessage,
 * and it is checked whether there is a mismatch between the description address and the packet source address.
 *
 * @receiver SsdpMessage to check
 * @param sourceAddress source address
 * @return true: if there is an invalid Location, such as a mismatch with the sender. false: otherwise
 */
internal fun SsdpMessage.hasInvalidLocation(sourceAddress: InetAddress): Boolean =
    (!hasValidLocation(sourceAddress)).also {
        if (it) Logger.w { "Location: $location is invalid from $sourceAddress" }
    }

private fun SsdpMessage.hasValidLocation(sourceAddress: InetAddress): Boolean {
    val location = location ?: return false
    if (!Http.isHttpUrl(location)) {
        return false
    }
    try {
        return sourceAddress == InetAddress.getByName(URL(location).host)
    } catch (ignored: IOException) {
    }
    return false
}

internal fun SsdpMessage.isNotUpnp(): Boolean {
    if (getHeader("X-TelepathyAddress.sony.com") != null) {
        // Sony telepathy service(urn:schemas-sony-com:service:X_Telepathy:1) send ssdp packet,
        // but it's location address refuses connection. Since this is not correct as UPnP, ignore it.
        Logger.v("ignore sony telepathy service")
        return true
    }
    return false
}
